Track global player rating+ranking using a Steam Leaderboard and the MultiElo algorithm

pelamfi opened the pull request:
- Add MultiElo algorithm and a test for it
- Add some support for the Steam leaderboards API
- Add component PilotRatingsUpdater to server side to compute ratings
- Transfer ratings to clients
- Add component on client PilotRatingUploader that uploads player's own rating to the leaderboard

To be done in upcoming PRs: visualize the player ratings and ranks in the main menu. Possibly show them also in the round ending score screen.

vvnurmi commented on AssaultWingCore/Stats/SteamLeaderboardService.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace AW2.Stats
+{
+  /// Game server API does not seem to have leaderboards:
```
Kind request to indent with 4 spaces as in existing cs files. :-)

pelamfi committed:
Introduce MultiElo multi player rating algorithm and tests

Implementation is based on examining this article and the related
Python repository:
https://towardsdatascience.com/developing-a-generalized-elo-rating-system-for-multiplayer-games-b9b495e87802

pelamfi committed:
SteamApiService: add SteamID getter

pelamfi committed:
Player: Add PilotRanking property to the Player

pelamfi committed:
LeaderboardEntryAndUgc: Introduce helper struct

pelamfi committed:
PilotRatingsUpdater: Introduce

- Start new players with rating of 1000
- Logging
- Allow calculating ratings even when players whose ratings are missing
and when there are only 2 players.

pelamfi committed:
PilotRatingUploader: Introduce

pelamfi committed:
UserControlledLogic: Create PilotRatingUploader

pelamfi committed:
AssaultWing: create SteamLeaderboardService

pelamfi committed:
NetworkEngine: GetSteamId

pelamfi committed:
NetworkEngineRaw: GetSteamId

pelamfi committed:
NetworkEngineSteam: GetSteamId

pelamfi committed:
ConnectionSteam: SteamId is CSteamID

pelamfi committed:
NetworkEngineSteam: Allow connecting multiple clients from same host

pelamfi committed:
PilotRanking: Introduce struct

pelamfi committed:
SteamLeaderboardService: Introduce

pelamfi committed:
Standings: Add FindForSpectator

pelamfi committed:
DedicatedServerLogic: Create and call PilotRatingsUpdater component

pelamfi committed:
MultiElo: Don't produce ratings below zero

pelamfi committed:
MultiEloTest: Don't produce ratings below zero

pelamfi committed:
MultiElo: Pass through the Score to output for debugging and logging purposes

pelamfi committed:
SerializationModeFlags: KeepLocalClientOwnedData

pelamfi committed:
SteamLeaderboardService: UploadCurrentPilotRanking return status boolean

pelamfi committed:
MessageHandlers: Support player rating transfer

pelamfi committed:
Player: support player rating transfer
